### PR TITLE
Clarify "full Unicode support"

### LINKED ISSUE #22: Support for UTF-16 surrogate pairs is not interoperable
~Unicode characters~ Surrogate pairs are not supported in .Net.  See example [here](http://regexstorm.net/tester?p=a.b&i=a%f0%90%84%81bc). 

The main goal of this specification as stated is to provide an interoperable subset of major Regex implementations (from the [abstract](https://www.ietf.org/archive/id/draft-ietf-jsonpath-iregexp-03.html#abstract), emphasis mine):

> This document specifies I-Regexp, a flavor of regular expressions that is limited in scope with the _goal of interoperation across many different regular-expression libraries_.

Mention of full unicode support needs to be ~removed~ downgraded to UTF-8 support.

- Introduction
    > I-Regexp supports the entire repertoire of Unicode characters.
- I-Regexp Syntax
    > In particular, full Unicode support is REQUIRED; the implementation MUST NOT limit itself to 7- or 8-bit character sets such as ASCII and MUST support the Unicode character property set in character classes.

Ref: https://github.com/jsonpath-standard/jsonpath-compliance-test-suite/pull/30/files#r1162157209

I'm not sure the reasoning behind adding this requirement.

### PR DESCRIPTION
**Pull request by cabo**
Close #22

### CODE REVIEW

File: draft-ietf-jsonpath-iregexp.md [hunk 1]
```diff
@@ -92,7 +92,8 @@ in this specification; the plural is "I-Regexps".
 {::boilerplate bcp14-tagged}
 
 The grammatical rules in this document are to be interpreted as ABNF,
-as described in {{-abnf}} and {{-abnf-cs}}.
+as described in {{-abnf}} and {{-abnf-cs}}, where the "characters" of
+{{Section 2.3 of -abnf}} are Unicode scalar values.
```
Comment (gregsdennis) [inline draft-ietf-jsonpath-iregexp.md:107]: Is Section 2.3 right?  Maybe I followed it wrong, but I get https://www.rfc-editor.org/rfc/rfc5234#section-2.3, which is titled "Termination Values."

**Review by gregsdennis — CHANGES_REQUESTED** [review]
I don't see how this clarifies what "full support" means.  The point of #22 is support of surrogate pairs.

File: draft-ietf-jsonpath-iregexp.md (hunk 1 above)
Comment (cabo) [reply draft-ietf-jsonpath-iregexp.md:107]: It's "Terminal Values", and the first paragraph explains that these used to be ASCII characters, and the new text explains that the terminal values are Unicode scalar values.

**Comment by cabo** [comment]
> I don't see how this clarifies what "full support" means. The point of #22 is support of surrogate pairs.

It clarifies that we don't expect "full Unicode support" in the sense of including every bell and whistle of any document of the Unicode Consortium ever written, but support for the features specified in this document.
(Surrogate pairs are not relevant for this document.)

**Comment by gregsdennis** [comment]
> Surrogate pairs are not relevant for this document.

This.  Where does the document say this?

**Comment by cabo** [comment]
> > Surrogate pairs are not relevant for this document.
> 
> This. Where does the document say this?

It also doesn't say that borscht recipes are not relevant for this document.

If you want to dig a bit deeper, search for "surrogate" in [XSD11-2] (or [XSD-2].

**Comment by gregsdennis** [comment]
> If you want to dig a bit deeper, search for "surrogate" in [XSD11-2] (or [XSD-2].

(I still don't understand why a spec essentially created for _JSON_ Path is referencing XML specs.)

I think it's worth explicitly mentioning that surrogate pairs aren't included, unless you're expecting every reader to be familiar with every nuance contained in every referenced document.

**Comment by cabo** [comment]
> > If you want to dig a bit deeper, search for "surrogate" in [XSD11-2] (or [XSD-2].
> 
> (I still don't understand why a spec essentially created for _JSON_ Path is referencing XML specs.)

Because they did their homework, so they are useful specs to reference.

> I think it's worth explicitly mentioning that surrogate pairs aren't included, unless you're expecting every reader to be familiar with every nuance contained in every referenced document.

"surrogate pairs aren't included" is not a meaningful way to say "hey, we are not inheriting the problems platforms with legacy 16-bit string models might have with Unicode handling".  I still can't believe that dotnet didn't do what ECMAScript did in 2015; are you sure about that?

**Comment by timbray** [comment]
> I think it's worth explicitly mentioning that surrogate pairs aren't included

Hmm. I can still see a case for forbidding surrogates but not just pairs. Remember that we say nothing about how the JSON text is actually encoded, just that we work on a JSON value. I don't think RFC8259 makes it impossible for surrogates (solo or in pairs) to appear in member names and string values, so I do see value in ruling that out.

Hmm, having said that, I looked at the draft, and especially "2.1.1 Syntax" and I don't see an assertion that a JSONPath expression MUST be a sequence of Unicode scalar values. It's perfectly possible I missed that but if I didn't, probably we should add that?

**Comment by cabo** [comment]
> > I think it's worth explicitly mentioning that surrogate pairs aren't included
> 
> Hmm. I can still see a case for forbidding surrogates but not just pairs. Remember that we say nothing about how the JSON text is actually encoded, just that we work on a JSON value. I don't think RFC8259 makes it impossible for surrogates (solo or in pairs) to appear in member names and string values, so I do see value in ruling that out.

Being more explicit about the character model (text string = sequence of Unicode Scalar Values) is a good thing.
Note that iregexp is not necessarily tied to JSON, so I'm assuming your comment is more about JSONPath, but a clarification about iregexp's character model would also be good.

> Hmm, having said that, I looked at the draft, and especially "2.1.1 Syntax" and I don't see an assertion that a JSONPath expression MUST be a sequence of Unicode scalar values. It's perfectly possible I missed that but if I didn't, probably we should add that?

That would be a comment on the other draft.

(The syntax of JSONPath expressions is described in ABNF, which this PR clarifies to be in terms of Unicode Scalar Values.  So there is no way to describe a JSONPath expression that is not a sequence of Unicode Scalar Values.  But, yes, being more explicit can help.)

**Comment by cabo** [comment]
[5d2ef98](https://github.com/ietf-wg-jsonpath/iregexp/pull/23/commits/5d2ef980b2be0392674d9551d4a261d9133204ce) made more explicit that both the iregexp strings and the strings iregexps operate on are built of Unicode scalar values.

**Comment by gregsdennis** [comment]
> Unicode scalar values

I had to look this up.  I found https://stackoverflow.com/a/48465266/878701.  Specifically (quoted from the spec):

> the set of Unicode scalar values consists of the ranges 0 to D7FF16 and E00016 to 10FFFF16, inclusive.

I don't understand how this translates to the hex values present in surrogate pairs (I had to look up the algorithms over a decade ago when I created my JSON parser, and any understanding has since left me.), so I don't know whether this includes the `\uD800\uDD01` present in the [proposed test](https://github.com/jsonpath-standard/jsonpath-compliance-test-suite/pull/30).

**Comment by cabo** [comment]
> > Unicode scalar values
> 
> I had to look this up. 

We have a definition over in JSONPath.  
I don't think we want to have another one here.

> I found https://stackoverflow.com/a/48465266/878701. Specifically (quoted from the spec):
> 
> > the set of Unicode scalar values consists of the ranges 0 to D7FF16 and E00016 to 10FFFF16, inclusive.
> 
> I don't understand how this translates to the hex values present in surrogate pairs (I had to look up the algorithms over a decade ago when I created my JSON parser, and any understanding has since left me.), so I don't know whether this includes the `\uD800\uDD01` present in the [proposed test](https://github.com/jsonpath-standard/jsonpath-compliance-test-suite/pull/30).

 D800 and DD01 are between D7FF and E000, so they are not Unicode scalar values.
However, "\uD800\uDD01" is just the JSON way for representing what more modern languages would express as "\u{10101}".  No surrogate pairs involved.

*** Aegean Numbers (Common)
𐄁: U+10101    1 AEGEAN WORD SEPARATOR DOT

**Comment by gregsdennis** [comment]
> D800 and DD01 are between D7FF and E000, so they are not Unicode scalar values.

Oh... It's a formatting error.  They had the `16` as a subscript indicating hex, and it pasted weird here.  Should be

> ranges 0 to D7FF<sub>16</sub> and E000<sub>16</sub> to 10FFFF<sub>16</sub>, inclusive

Now I see it.

---

> We have a definition over in JSONPath.
> I don't think we want to have another one here.

But JSON Path depends on this.  If anything, we'd want it here.

**Comment by cabo** [comment]
> But JSON Path depends on this. If anything, we'd want it here.

I don't think every specification that uses Unicode has to repeat all of the Unicode specifications.

First google hit for me leads to:
http://www.unicode.org/glossary/#unicode_scalar_value

Grüße, Carsten

File: draft-ietf-jsonpath-iregexp.md (hunk 1 above)
Comment (gregsdennis) [reply draft-ietf-jsonpath-iregexp.md:107]: okay.  (not sure why I can't mark this resolved)

**Review by gregsdennis — APPROVED** [review]

**Review by timbray — APPROVED** [review]

**Pull request merged**